### ISSUE TITLE
iOS support

### ISSUE THREAD

**Issue by ribtoks** [ihhub/fheroes2#7043]
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.

### Describe the problem requiring a solution

I think this game would be perfect on an iPad, but currently there's no way to launch it...

### Describe the possible solution

You could provide an .ipa file (like the VCMI project) for side-loading. Ideally it will of course be a paid app on the AppStore (so that you could receive some money)

**Comment by ihhub** [comment]
Hi @ribtoks , as of now we don't have plans to release fheroes2 on iOS due to lack of resources. Android release was necessary due to openness of the platform and amount if potential players. iOS is much more restrictive in this matter and also we have no copyright rights to ask any money for the engine. So free is the only option.

As a conclusion, we are not saying a no but for now we can't focus on it unless someone steps in and helps with porting to iOS.

**Comment by ribtoks** [comment]
Thank you for your response @ihhub . I hope the situation with iOS will change in the future and I wish best for fheroes2 project!

**Comment by dbratcher** [comment]
@ihhub I'm an iOS engineer. I'm new to this project but a long time heroes fan. Just curious, if I wanted to explore adding to iOS support does the project have a general idea on how iOS support would be added? I'm wondering if looking at the Android implementation and straight porting relevant code or investigating however Android reused code would be useful.

**Comment by oleg-derevenetz** [comment]
@ihhub

> and also we have no copyright rights to ask any money for the engine

Technically, the GPL does not prevent you from asking for money for the engine :)

@dbratcher

> Just curious, if I wanted to explore adding to iOS support does the project have a general idea on how iOS support would be added? I'm wondering if looking at the Android implementation and straight porting relevant code or investigating however Android reused code would be useful.

Most of Android code is related to the "companion" toolset app which is used to provide the assets of the original game and manage the save files (since on modern Android versions you can't just read contents of some public directory on SD card or tell a user to copy assets to the app data directory - exactly the same as on iOS by the way). The only our Android class that is directly related to the game engine itself is [this one](https://github.com/ihhub/fheroes2/blob/master/android/app/src/main/java/org/fheroes2/GameActivity.java), and it's relatively straightforward.

Here is a [checklist, as well as a kind of retrospective](https://github.com/ihhub/fheroes2/issues/6008) of porting this game to Android, many things that are necessary for the game to work on handheld devices with touchscreen (such as emulation of right mouse button) are already implemented. SDL in addition to Android also supports iOS, so technically everything is possible, I believe.

**Comment by dbratcher** [comment]
Awesome thank you for all of this, I'll start checking it out.

**Comment by ihhub** [comment]
Hi @dbratcher , also if you think of an easier way for end-users to install ipa file please feel to suggest.

**Comment by oleg-derevenetz** [comment]
> also if you think of an easier way for end-users to install ipa file please feel to suggest.

AFAIK there is no way to install an ipa file without jailbreaking.

**Comment by ribtoks** [comment]
> there is no way to install an ipa file without jailbreaking.
@dbratcher @oleg-derevenetz  this is not true and it's possible to install - see [VCMI instructions](https://wiki.vcmi.eu/Installation_on_iOS). In fact, I have VCMI installed on iPad, signed by my developer certificate and it works.

**Comment by oleg-derevenetz** [comment]
> signed by my developer certificate

So end user needs to have the developer certificate which means that we are no longer actually talking about the "end user" :) People with developer certificate can just build it from source using Xcode (and don't use ipa at all), but what is the percentage of such people among the total number of iOS users?

**Comment by ribtoks** [comment]
> what is the percentage of such people among the total number of iOS users?

This is a valid point. I would just note that a) it will be at least some way to use it on iOS and b) it's good to start somewhere

**Comment by ribtoks** [comment]
> People with developer certificate can just build it from source using Xcode (and don't use ipa at all)

Can I build it today using XCode and install on an iOS device?

**Comment by oleg-derevenetz** [comment]
> Can I build it today using XCode and install on an iOS device?

No, not today :) At a minimum, some refinement will be required, although, I hope, not much (compared to desktop -> Android), since some changes (related to handheld devices in general) are universal.

**Comment by dbratcher** [comment]
Teaser of what I have running so far on iPad 😂![image](https://user-images.githubusercontent.com/664567/235333120-7848793e-b421-4757-a493-cdff63d0f38e.jpeg)

**Comment by dbratcher** [comment]
I'm working to add support here: https://github.com/dbratcher/fheroes2/tree/ios_support

**Comment by dbratcher** [comment]
Very very hacky... but working 😁

 ![image](https://user-images.githubusercontent.com/664567/235335622-5ce39922-4d22-42b1-bddd-3cf3d198f495.jpeg)

**Comment by oleg-derevenetz** [comment]
> Very very hacky... but working 😁

I suppose a kind of "launcher" UI will be needed on iOS as well to handle issues with original game assets in some user-friendly way. Also I'm not sure how iOS works with MIDI (macOS has native support for MIDI playback).

**Comment by ihhub** [comment]
Hi @dbratcher , once you are ready please create a pull request. I believe we can do a launcher later. Also are you able to build it using GitHub Actions?

**Comment by dbratcher** [comment]
@ihhub I'll clean up the code first, I'd be a bit embarrassed to make a PR in its current state 😅 things on my radar:
- structure code to be more dry
- improve loading logic to load all relevant assets (only loads one map at the moment)
- fix introduced warnings
- properly gate code so it doesn't break other platforms
- run GitHub actions and fix any issues
- create a PR with documentation

Long term efforts
- add support to data is missing screen to import assets from Apple Files and have instructions
- go through Android platform checks and see if similar changes are needed for iOS
- go through Android development checklist and see if similar work is needed
- determine if it's worth putting on the App Store (could we bundle the demo assets with it?)
- maybe add a donate button for support

**Comment by ihhub** [comment]
Hi @dbratcher ,
> improve loading logic to load all relevant assets (only loads one map at the moment)
Are you sure about it? Don't you use a demo version of the game?

**Comment by dbratcher** [comment]
Yea I'm sure, if you look at my code it's real hacky. I use a good old game version

**Comment by ribtoks** [comment]
Hey @dbratcher No pressure really, just curious if you had any luck with your proof-of-concept? Do you think what you did is somewhat reproducible so that for example I can try it "on my computer"?

**Comment by drewbratcher** [comment]
@ribtoks unfortunately I haven't touched it since that first effort awhile back. If you wanted to try and replicate what I did you could checkout [my branch](https://github.com/dbratcher/fheroes2/tree/ios_support) and then look at how I added the required data files in [this commit](https://github.com/dbratcher/fheroes2/commit/05ff1bd1e244490f387a7a9a61b38dd0731c659f). The data files are ignored by the git repo so they're not checked in and you'll need to get them from a game install. That commit also shows the very hacky code I added to load the data files, I don't know c++ and had a lot of trouble with it. You have to make sure the files are copied over to the app bundle as well I believe.

Let me know if you try anything out and get it working 😁 If not, I do still plan on making this easier to set up and use when I get some more time.

**Comment by peterpaulis** [comment]
+1 would love to play this on iPad similar as H3 VCMI

**Comment by oldfoz** [comment]
Being able to play on iOS devices is fantastic

VCMI project forced me to handle how to operate with AltStore to play H3,
and I'm looking forward now to the timeless classic H2 :)

Thank you for your contribution!

**Comment by dbratcher** [comment]
hey all, really sorry I haven't been able to spend more time on this. I put what I had in a [pull request here](https://github.com/ihhub/fheroes2/pull/8407). Unfortunately I can't spend much time on it right now, but maybe it can help someone else if they want to clean it up for possible future iOS support

**Comment by peterpaulis** [comment]
any change year and half after? would love to play on iPad as VCMI

**Comment by ihhub** [comment]
> any change year and half after? would love to play on iPad as VCMI

You could help the team to make it happen :)

**Comment by ihhub** [comment]
The initial implementation was done in #8407. I am closing this issue and will open another one with a list of proper steps what to do next.

**Comment by kambala-decapitator** [comment]
FYI we tried to get VCMI into AppStore and hit a major block so far: the "user must add external files to the app to be able to use it" approach is allowed only for apps classified as emulators, and they don't care that those files aren't executable code.

**Comment by ihhub** [comment]
> FYI we tried to get VCMI into AppStore and hit a major block so far: the "user must add external files to the app to be able to use it" approach is allowed only for apps classified as emulators, and they don't care that those files aren't executable code.

Thanks! We expect this. To reach this state we need to first make a iOS version to at least be installed by users outside Apple Store. Then, an app can actually download a demo version of the game. Also, it is a must that the resources should be available via Files app on iPhone.

**Comment by ribtoks** [comment]
@ihhub amazing! but where can I see the instructions so I could try this? I couldn't see anything about iOS in docs.

**Comment by ihhub** [comment]
> [@ihhub](https://github.com/ihhub) amazing! but where can I see the instructions so I could try this? I couldn't see anything about iOS in docs.

The instructions are exactly in docs: https://github.com/ihhub/fheroes2/blob/master/docs/DEVELOPMENT.md

**Comment by peterpaulis** [comment]
Could somebody please create an .ipa, that i could self sign using appsigner (https://www.iosappsigner.com/), similar to vcmi?

**Comment by CurishDakoma** [comment]
just leaving it here - https://developer.apple.com/support/dma-and-apps-in-the-eu/ & https://developer.apple.com/support/dma-and-apps-in-the-eu/#ios-app-eu

it is no longer necessary to have app in apple store.
